src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/8.diff.sql

```diff
@@ -692,3 +692,234 @@ AS
+--     @referenceSearchParams
+--         * Extracted reference search params
+--     @tokenSearchParams
+--         * Extracted token search params
+--     @tokenTextSearchParams
+--         * The text representation of extracted token search params
+--     @stringSearchParams
+--         * Extracted string search params
+--     @numberSearchParams
+--         * Extracted number search params
+--     @quantitySearchParams
+--         * Extracted quantity search params
+--     @uriSearchParams
+--         * Extracted URI search params
+--     @dateTimeSearchParms
+--         * Extracted datetime search params
+--     @referenceTokenCompositeSearchParams
+--         * Extracted reference$token search params
+--     @tokenTokenCompositeSearchParams
+--         * Extracted token$token tokensearch params
+--     @tokenDateTimeCompositeSearchParams
+--         * Extracted token$datetime search params
+--     @tokenQuantityCompositeSearchParams
+--         * Extracted token$quantity search params
+--     @tokenStringCompositeSearchParams
+--         * Extracted token$string search params
+--     @tokenNumberNumberCompositeSearchParams
+--         * Extracted token$number$number search params
+--
+CREATE PROCEDURE dbo.ReindexResource
+    @resourceTypeId smallint,
+    @resourceId varchar(64),
+    @eTag int = NULL,
+    @searchParamHash varchar(64),
+    @resourceWriteClaims dbo.ResourceWriteClaimTableType_1 READONLY,
+    @compartmentAssignments dbo.CompartmentAssignmentTableType_1 READONLY,
+    @referenceSearchParams dbo.ReferenceSearchParamTableType_2 READONLY,
+    @tokenSearchParams dbo.TokenSearchParamTableType_1 READONLY,
+    @tokenTextSearchParams dbo.TokenTextTableType_1 READONLY,
+    @stringSearchParams dbo.StringSearchParamTableType_1 READONLY,
+    @numberSearchParams dbo.NumberSearchParamTableType_1 READONLY,
+    @quantitySearchParams dbo.QuantitySearchParamTableType_1 READONLY,
+    @uriSearchParams dbo.UriSearchParamTableType_1 READONLY,
+    @dateTimeSearchParms dbo.DateTimeSearchParamTableType_1 READONLY,
+    @referenceTokenCompositeSearchParams dbo.ReferenceTokenCompositeSearchParamTableType_2 READONLY,
+    @tokenTokenCompositeSearchParams dbo.TokenTokenCompositeSearchParamTableType_1 READONLY,
+    @tokenDateTimeCompositeSearchParams dbo.TokenDateTimeCompositeSearchParamTableType_1 READONLY,
+    @tokenQuantityCompositeSearchParams dbo.TokenQuantityCompositeSearchParamTableType_1 READONLY,
+    @tokenStringCompositeSearchParams dbo.TokenStringCompositeSearchParamTableType_1 READONLY,
+    @tokenNumberNumberCompositeSearchParams dbo.TokenNumberNumberCompositeSearchParamTableType_1 READONLY
+AS
+    SET NOCOUNT ON
+
+    SET XACT_ABORT ON
+    BEGIN TRANSACTION
+
+    DECLARE @resourceSurrogateId bigint
+    DECLARE @version bigint
+
+    -- This should place a range lock on a row in the IX_Resource_ResourceTypeId_ResourceId nonclustered filtered index
+    -- TODO: Should we be locking all index tables too?
```

Comment:
@johnstairs this lock is the same one that we are using in the `UpsertResource` stored proc. 

To confirm my understanding, this is only locking certain access to the `dbo.Resource` table, right? If so, is it alright that we are not blocking read/write access to all the index tables during an upsert or reindex operation?

Reply:
All reads and writes that you do within a transaction will place different kinds of locks.

Since we are using RCSI, reads will not be blocked by this operation. For writes, concurrent calls to this sproc and  UpsertResource would block each other if accessing the same resource, but that should be by design. Updating separate resources should be fine.

Reply:
Thanks for the reply!

I felt like I still needed a refresher on this, so I read through [this doc](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/sql/snapshot-isolation-in-sql-server). Jotting down my takeaways here for my future self:

- Read Committed Snapshot Isolation (RCSI) ensures that data read within a transaction will never reflect changes made by other simultaneous transactions
- With RSCI, the data row versions are recorded at the beginning of a transaction, and if a transaction attempts to commit an update to a row that was changed after the transaction began, the transaction is rolled back, and an error is raised
- Lock hints like `UPDLOCK` can reduce the chance of update conflicts in long running transactions by blocking other transactions from attempting to update a specific set of rows before the transaction completes

The stored proc above begins a new transaction which has RSCI enabled. The lock is used to block other transactions from updating the row in `dbo.Resource` for the resource being reindexed. In theory, the data in the index tables relevant to this resource could be modified over the course of the transaction. If that were the case, the transaction would fail to commit, and an error would be returned.

@johnstairs please let me know if I have misunderstood something here
@feordin it sounds like this would be another similar scenario where we would like to reattempt to reindex a resource